File: fushi/test/pages/popup_mining_hidden_dictionary_test.js
// TODO-865 behavior test (BUG-419 sibling): a dictionary disabled in 词典管理
// (its show/hide switch off → added to window.hiddenDictionaryNames by the host)
// must NOT have its definitions written into the Anki mining payload.
//
// Root cause: term dictionaries stay registered in the native FFI engine even when
// hidden (AppModel.bucketDictPaths keeps hidden term dicts in the bucket because
// filtering is meant to happen at render time). BUG-419 only patched the visible
// lookup popup (createGlossarySectionWrapper). The mining field assembly path
// (constructGlossaryHtml / constructSingleGlossaryHtml, consumed by buildMinePayload)
// is independent and previously had NO hidden filter, so a disabled dictionary's
// glossary still ended up in the card's glossary field.
//
// This test EXECUTES the real popup.js constructGlossaryHtml / constructSingleGlossaryHtml
// against a minimal fake DOM and asserts hidden dictionaries are excluded from the
// mining payload (and that an entry whose only glossary is hidden yields an empty
// <ol></ol> / {} — no empty card content). Reverting the fix turns this red.
//
// Run: node fushi/test/pages/popup_mining_hidden_dictionary_test.js
// (also driven from popup_mining_hidden_dictionary_test.dart so it executes inside
//  `flutter test`).

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const popupPath = path.resolve(__dirname, '../../assets/popup/popup.js');
const source = fs.readFileSync(popupPath, 'utf8');

function makeElement(tag) {
  return {
    tagName: (tag || 'div').toUpperCase(),
    className: '',
    id: '',
    textContent: '',
    innerHTML: '',
    style: {},
    dataset: {},
    children: [],
    attributes: [],
    classList: {
      _set: new Set(),
      add(name) { this._set.add(name); },
      remove(name) { this._set.delete(name); },
      contains(name) { return this._set.has(name); },
    },
    appendChild(child) { this.children.push(child); return child; },
    append(...nodes) { this.children.push(...nodes); },
    setAttribute() {},
    getAttribute() { return null; },
    hasAttribute() { return false; },
    removeAttribute() {},
    addEventListener() {},
    querySelectorAll() { return []; },
    querySelector() { return null; },
    closest() { return null; },
    get firstChild() { return this.children.length ? this.children[0] : null; },
  };
}

function makeSandbox(hiddenNames) {
  const documentObj = {
    documentElement: { style: {}, classList: makeElement().classList },
    head: { appendChild() {} },
    body: makeElement('body'),
    getElementById() { return null; },
    querySelector() { return null; },
    querySelectorAll() { return []; },
    createElement(tag) { return makeElement(tag); },
    createTextNode(text) { const n = makeElement('#text'); n.textContent = text; return n; },
    addEventListener() {},
  };

  const windowObj = {
    audioSources: [],
    needsAudio: false,
    lookupEntries: [],
    dictionaryStyles: {},
    hiddenDictionaryNames: hiddenNames || [],
    collapsedDictionaryNames: [],
    compactGlossariesAnki: false,
    flutter_inappwebview: { callHandler() { return Promise.resolve(false); } },
    getSelection() { return { toString() { return ''; } }; },
  };
  documentObj.defaultView = windowObj;

  const sandbox = {
    Node: { TEXT_NODE: 3, ELEMENT_NODE: 1 },
    Date, Math, URL, JSON, RegExp, Set, Map, Object, Array, console,
    performance: { now() { return 0; } },
    setTimeout, clearTimeout,
    DOMParser: class { parseFromString() { return { body: makeElement('body'), querySelectorAll() { return []; } }; } },
    document: documentObj,
    window: windowObj,
    getComputedStyle() { return {}; },
  };
  sandbox.globalThis = sandbox;
  return sandbox;
}

function loadPopup(hiddenNames, entry) {
  const sandbox = makeSandbox(hiddenNames);
  vm.createContext(sandbox);
  const exported = source + `
    ;window.lookupEntries = [${JSON.stringify(entry)}];
    ;window.__test = {
      multi: function() { return constructGlossaryHtml(0); },
      single: function() { return constructSingleGlossaryHtml(0); },
    };
  `;
  vm.runInContext(exported, sandbox, { filename: 'popup.js' });
  return sandbox;
}

function gloss(dictionary) {
  // content is a plain (non-JSON, non-HTML) string so renderStructuredContent
  // treats it as text; the data-dictionary marker is what we assert on.
  return { dictionary: dictionary, content: 'def-' + dictionary, definitionTags: '', termTags: '' };
}

function entryWith(dictNames) {
  return {
    expression: '猫', reading: 'ねこ',
    glossaries: dictNames.map(gloss),
    frequencies: [], pitches: [],
  };
}

(function run() {
  // Test 1: with no hidden dicts, both dictionaries appear in both payload shapes.
  {
    const sb = loadPopup([], entryWith(['JMdict', 'Daijirin']));
    const multi = sb.window.__test.multi();
    assert.ok(multi.includes('data-dictionary="JMdict"'),
      'visible JMdict must be in multi glossary; got ' + multi);
    assert.ok(multi.includes('data-dictionary="Daijirin"'),
      'visible Daijirin must be in multi glossary; got ' + multi);

    const single = sb.window.__test.single();
    assert.deepStrictEqual(Object.keys(single).sort(), ['Daijirin', 'JMdict'],
      'with no hidden dicts both must be in single glossaries; got '
        + JSON.stringify(Object.keys(single)));
  }

  // Test 2: a hidden dictionary is excluded from the mining payload; visible survives.
  {
    const sb = loadPopup(['Daijirin'], entryWith(['JMdict', 'Daijirin']));
    const multi = sb.window.__test.multi();
    assert.ok(multi.includes('data-dictionary="JMdict"'),
      'visible JMdict must remain in multi glossary; got ' + multi);
    assert.ok(!multi.includes('data-dictionary="Daijirin"'),
      'disabled Daijirin must NOT appear in the mining multi glossary; got ' + multi);

    const single = sb.window.__test.single();
    assert.deepStrictEqual(Object.keys(single), ['JMdict'],
      'disabled dictionary must be filtered from single glossaries; got '
        + JSON.stringify(Object.keys(single)));
    assert.strictEqual(single['Daijirin'], undefined,
      'the hidden dictionary must not be a key in single glossaries');
  }

  // Test 3: an entry whose only glossary is hidden yields empty payloads
  // (empty <ol></ol> with no <li>, and {} for single) — no empty card content.
  {
    const sb = loadPopup(['JMdict'], entryWith(['JMdict']));
    const multi = sb.window.__test.multi();
    assert.ok(!multi.includes('<li'),
      'an all-hidden entry must produce no glossary list items; got ' + multi);
    assert.ok(!multi.includes('data-dictionary="JMdict"'),
      'the hidden dictionary must not surface at all; got ' + multi);

    const single = sb.window.__test.single();
    assert.strictEqual(Object.keys(single).length, 0,
      'an all-hidden entry must produce an empty single-glossary map; got '
        + JSON.stringify(single));
  }

  console.log('popup_mining_hidden_dictionary_test.js: all assertions passed');
})();

File: fushi/test/pages/popup_mining_hidden_dictionary_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// TODO-865 (BUG-419 sibling): a dictionary disabled in 词典管理 (its show/hide
/// switch off) must not have its definitions written into the Anki mining payload.
///
/// 根因：term 词典即使被隐藏（disabled）也仍注册进 FFI 引擎——`AppModel.bucketDictPaths`
/// 的 term 分支故意不按 hidden 过滤（注释「term 在渲染期按 hidden 过滤，故隐藏仍进桶」），
/// 因此 `entry.glossaries` 仍含被禁用词典的义项。BUG-419 只堵了可视查词弹窗
/// （`createGlossarySectionWrapper`）；制卡走另一条独立的字段组装路径
/// （`constructGlossaryHtml` / `constructSingleGlossaryHtml`，由 `buildMinePayload` 消费），
/// 此前没有消费 `window.hiddenDictionaryNames`，把隐藏词典义项原样塞进卡片 glossary 字段。
/// 修复：两函数各加一行 `hiddenDictionaryNames.includes(g.dictionary)) return;`，与
/// `createGlossarySectionWrapper`（popup.js:1679-1682）完全同款、同源谓词。
///
/// 两层守护：
/// ① 行为级——用 Node 真执行 popup.js 的 `constructGlossaryHtml` /
///    `constructSingleGlossaryHtml`，断言隐藏词典被排除、且仅隐藏词典的词条产出空
///    `<ol></ol>` / `{}`。无 node 时 skip。
/// ② 源码级——静态扫描 popup.js 两函数体内各含 hidden 过滤行（即便无 node 也守得住）。
void main() {
  test(
    'mining payload excludes disabled dictionaries '
    '(executes constructGlossaryHtml / constructSingleGlossaryHtml via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest = File(
        'test/pages/popup_mining_hidden_dictionary_test.js',
      );
      expect(
        jsTest.existsSync(),
        isTrue,
        reason: 'behavior harness ${jsTest.path} must exist',
      );

      final ProcessResult result = await Process.run(
        nodeExe,
        <String>[jsTest.path],
        workingDirectory: Directory.current.path,
      );

      expect(
        result.exitCode,
        0,
        reason:
            'popup mining hidden-dictionary filter JS behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(
        result.stdout.toString(),
        contains('all assertions passed'),
        reason: 'behavior harness must reach its success marker',
      );
    },
  );

  test(
    'popup.js drops hidden dictionaries in both mining glossary builders',
    () {
      final String js = File('assets/popup/popup.js').readAsStringSync();

      // constructSingleGlossaryHtml must consult the hidden names list and skip
      // those dictionaries inside its glossary forEach.
      _assertHiddenFilterInFunction(js, 'constructSingleGlossaryHtml');

      // constructGlossaryHtml must do the same.
      _assertHiddenFilterInFunction(js, 'constructGlossaryHtml');
    },
  );
}

/// Asserts that the body of [functionName] (between its `function <name>(` and the
/// next top-level `function ` declaration) reads `window.hiddenDictionaryNames` and
/// skips hidden dictionaries via `hiddenDictionaryNames.includes(g.dictionary)) return;`
/// inside `entry.glossaries.forEach(`.
void _assertHiddenFilterInFunction(String js, String functionName) {
  final int start = js.indexOf('function $functionName(');
  expect(start, greaterThanOrEqualTo(0), reason: '$functionName must exist');

  final int next = js.indexOf('\nfunction ', start + 1);
  final String body =
      next >= 0 ? js.substring(start, next) : js.substring(start);

  expect(
    body.contains('window.hiddenDictionaryNames'),
    isTrue,
    reason: '$functionName must read window.hiddenDictionaryNames',
  );

  final int forEach = body.indexOf('entry.glossaries.forEach(');
  final int skip =
      body.indexOf('hiddenDictionaryNames.includes(g.dictionary)) return;');
  expect(forEach, greaterThanOrEqualTo(0),
      reason: '$functionName must iterate entry.glossaries');
  expect(skip, greaterThan(forEach),
      reason: 'hidden dictionaries must be skipped inside the glossary forEach '
          'of $functionName');
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}
